[FSDP2] Used `ReduceOp.AVG` if fp32 reduce-scatter

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120919
* #120910
* #120238

This PR uses `ncclAvg` op (via `ReduceOp.AVG`) if doing fp32 reduce-scatter. This allows the division by world size to happen in the reduce-scatter kernel itself, which seems to save extra memory read/write for dividing. This yields ~1.5% speedup on the Llama-7B workload (and makes per-parameter FSDP faster than flat-parameter FSDP 😅 ).

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
[FSDP2] Changed to only post-divide if fp32 reduce

[ghstack-poisoned]

wanchaol commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -165,18 +164,24 @@ def _init_grad_divide_factors(self):
+        if self._reduce_dtype == torch.float32:
+            # Only divide after reduction since fp32 has sufficient precision
+            self._grad_predivide_factor: float = 1.0
+            self._grad_postdivide_factor: float = data_parallel_world_size
```
I actually keep forgetting why we want to do the "manual" division after reduce_scatter, instead of using `ReduceOp.AVG` when calling reduce_scatter?

awgu replied:
That is a good point. I tested on torchtrain Llama-7B, and `ReduceOp.AVG` is quite a bit faster I think since we avoid extra memory read/write for the div (as NCCL does it as it is reducing).

I updated the PR with this behavior. I am very interested to understand if we can do this change as well for bf16 reduce-scatter.

Commit:
Update on "[FSDP2] Changed to post-divide-only if fp32 reduce"


This PR makes the fp32 reduce-scatter only divide by world size after the reduce-scatter instead of splitting into a pre-divide and post-divide. This is the same behavior as DDP. This change helps save half the `div_` kernels (~1% speedup on Llama-7B 😆 ).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Changed to post-divide-only if fp32 reduce"


This PR makes the fp32 reduce-scatter only divide by world size after the reduce-scatter instead of splitting into a pre-divide and post-divide. This is the same behavior as DDP. This change helps save half the `div_` kernels (~1% speedup on Llama-7B 😆 ).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Used `ReduceOp.AVG` if fp32 reduce-scatter"


This PR makes the fp32 reduce-scatter only divide by world size after the reduce-scatter instead of splitting into a pre-divide and post-divide. This is the same behavior as DDP. This change helps save half the `div_` kernels (~1% speedup on Llama-7B 😆 ).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Used `ReduceOp.AVG` if fp32 reduce-scatter"


This PR uses `ncclAvg` op (via `ReduceOp.AVG`) if doing fp32 reduce-scatter. This allows the division by world size to happen in the reduce-scatter kernel itself, which seems to save extra memory read/write for dividing. This yields ~1.5% speedup on the Llama-7B workload (and makes per-parameter FSDP faster than flat-parameter FSDP 😅 ).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Used `ReduceOp.AVG` if fp32 reduce-scatter"


This PR uses `ncclAvg` op (via `ReduceOp.AVG`) if doing fp32 reduce-scatter. This allows the division by world size to happen in the reduce-scatter kernel itself, which seems to save extra memory read/write for dividing. This yields ~1.5% speedup on the Llama-7B workload (and makes per-parameter FSDP faster than flat-parameter FSDP 😅 ).

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

awgu commented on torch/distributed/_composable/fsdp/_fsdp_collectives.py:
```diff
@@ -197,6 +195,23 @@ def foreach_reduce_scatter_copy_in(
+    if divide_factors:
+        predivide_factor, postdivide_factor = divide_factors
+        _div_if_needed(input, predivide_factor)
+        dist.reduce_scatter_tensor(output, input, group=group)
```
`premul_sum` looks useful here! I think we can try to use it when we have the plumbing up to Python (IIUC, not yet?).

> Also btw, can you remind me why we are using dist.* instead of funcol here?

I think it is just legacy. Is the idea that, since we are allocating a new reduce-scatter output anyway, then there is no difference (e.g. no extra copies/allocations) in eager mode whether we use `dist.reduce_scatter_tensor` vs. `funcol.reduce_scatter_tensor`?

yifuwang replied:
> I think it is just legacy. Is the idea that, since we are allocating a new reduce-scatter output anyway, then there is no difference (e.g. no extra copies/allocations) in eager mode whether we use dist.reduce_scatter_tensor vs. funcol.reduce_scatter_tensor?

Ah that's a good point. You are right, `funcol.reduce_scatter_tensor` would perform similarly, but only if we allocated the output just-in-time in eager mode.

wanchaol replied:
I think in longer term we should consider to let funcol to accept `out` argument, this is similar behave to aten ops, and allow us to use funcol inside `dist.*` collective

yifuwang replied:
Yup this is totally doable. We can have out-variants funcols in eager, which get functionalized by AOTAutograd (we need to add functionalization glue for it I think. Not sure if the recent auto functionalization change changed this) then re-inplaced in the post-grad fx pass.

wanchaol reviewed (approved):
lgtm!

awgu commented:
`linux-binary-manywheel / manywheel-py3_8-cuda11_8-test / test (push)` failure unrelated:
```
RuntimeError: cuDNN version incompatibility: PyTorch was compiled  against (8, 9, 7) but found runtime version (8, 7, 0). PyTorch already comes bundled with cuDNN. One option to resolving this error is to ensure PyTorch can find the bundled cuDNN. one possibility is that there is a conflicting cuDNN in LD_LIBRARY_PATH.
```

awgu commented:
@pytorchbot merge -i
